### ISSUE TITLE
`ttir.embedding_backward` leaves rows unwritten when num_weights is not a multiple of 32 - reads garbage

### ISSUE THREAD

**Issue by agobeljicTT** [tenstorrent/tt-mlir#9220]
## Summary

`ttir.embedding_backward` leaves part of its result unwritten when `num_weights` is not a
whole number of 32-row tiles. Those rows come back holding whatever was in the buffer, so
rows that no index selected — which must be exactly zero — read back as garbage. Silent wrong
results: no error, no warning. Every multiple of 32 is clean.

BERT's 30522 and GPT-2's 50257 are both in this class.

## Repro

100 rows of 64, 8 indices, all of them 0, every gradient element 1.0. The only correct result
is row 0 = 8.0 and every other row = 0.0.

```mlir
module {
  func.func @main() -> tensor<100x64xbf16> {
    %indices = "ttir.constant"() <{value = dense<0> : tensor<1x8xsi32>}> : () -> tensor<1x8xsi32>
    %weight = "ttir.zeros"() <{shape = array<i32: 100, 64>}> : () -> tensor<100x64xbf16>
    %grad = "ttir.full"() <{shape = array<i32: 1, 8, 64>, fill_value = 1.0 : f32}> : () -> tensor<1x8x64xbf16>
    %r = "ttir.embedding_backward"(%indices, %weight, %grad) : (tensor<1x8xsi32>, tensor<100x64xbf16>, tensor<1x8x64xbf16>) -> tensor<100x64xbf16>
    return %r : tensor<100x64xbf16>
  }
}
```

```
ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=<sysdesc>" -o repro_ttnn.mlir repro.mlir
ttmlir-translate --ttnn-to-flatbuffer -o repro.ttnn repro_ttnn.mlir
ttrt run repro.ttnn --save-artifacts
```

Row 0 is correct (8.0). Rows in the trailing `100 % 32` come back holding uninitialized memory:

```
nonzero rows: [0, 96, 97, 99]        # only row 0 should be nonzero
row 96: first4=[0.0737, 1.1328, -0.208, 1.6172]    max=2.29e37   nonzero=64/64
row 97: first4=[-0.9023, -1.1719, -0.6523, -0.6445] max=1.60e38  nonzero=64/64
row 99: first4=[-1.2109, -0.2891, 1.4453, 0.418]   max=1.38e38   nonzero=62/64
```

Magnitudes up to 1e38 out of a computation whose only inputs are 0 and 1.0 — this is a read
of memory the kernel never wrote, not an arithmetic result. Row 98 read back zero on this
particular run, which is the same residue dependence described below, not a partial fix.

## Also measured through PyTorch

Through the PyTorch frontend (tt-kurbla, `aten::embedding_dense_backward`), bf16, hidden 64,
indices drawn from rows 0-7 only, so every row from 8 up must be zero:

| `num_weights` | rows that came back nonzero |
|---|---|
| 100 | 96, 97, 98, 99 (cold buffer) — 88, 89, 90, 91 after a large write to the same table |
| 120 | 16 rows from 104 |
| 129 | 128 (intermittent) |
| 200 | 192-199 |
| 64, 96, 128, 160, 256, 1000 | none |

Against an f32 reference the corrupt rows are off by whole row magnitudes (max abs error 2.98
where bf16 rounding accounts for 0.03), so this is not an accuracy question.

Which rows come back dirty, and whether the corruption is visible at all, depends on what last
held that buffer — a run whose predecessor left zeros there looks clean. `num_weights = 100`
reproduced on every attempt.

## Cause

Not diagnosed below the op — what follows is where I would start looking, not a finding.

`EmbeddingBackwardOpConversionPattern` (`TTIRToTTNN.cpp`) pads the *indices* and the
*gradient* out to a tile boundary before handing them to `ttnn.embedding_bw`, but nothing
pads or clears the result table. If the kernel only writes the rows it accumulated into and
leaves the rest of the final tile alone, that matches what comes back.

The rows affected are not always the last `num_weights % 32`, though: the same op reached
through PyTorch at hidden 64 dirtied rows 88-91 of a 100-row table rather than 96-99. So the
region left unwritten depends on more than the vocabulary remainder.

Nothing above this op can work around it: the affected rows are inside the vocabulary, so a
caller cannot tell a corrupt row from a legitimately updated one.

## Impact

Any embedding trained with a vocabulary that is not a multiple of 32 — BERT's 30522, GPT-2's
50257 — takes garbage into its weight gradient, so the optimizer writes it into rows for
tokens that never appeared in the batch, on every step. Vocabularies that are already
tile-aligned (Llama's 32000 and 128256, Qwen's 151936) are unaffected, which is likely why
this has gone unnoticed.

## Environment

- Blackhole, single chip.
- tt-mlir at 6d27eed, which includes #9163 ("Type embedding_bw's result as the 4D gradient
  tt-metal returns"). Not checked against older revisions.
- bf16 gradient; f32 behaves the same, having been rounded to bf16 by the kernel.
